### ISSUE TITLE
Turn on TS noUncheckedIndexAccess

### ISSUE THREAD

**Issue by mitschabaude** [o1-labs/o1js#1518]

**Comment by harrysolovay** [comment]
I don't think it would be worthwhile to refactor all of this code to check for index existence. I think a better approach will be to apply non-null assertions, and––from here on out––take advantage of `noUncheckedIndexAccess`. @mitschabaude want to assign this to me?

**Comment by mitschabaude** [comment]
@harrysolovay if i understand you correctly, I agree: in most cases the initial fix should be something like `assert(item !== undefined)` before using the item that was obtained with an index access -- instead of refactoring the array to a tuple type / checking that the key is in the object.

I'm happy to see you having a go at this. Since it's a big refactor, it could also be done iteratively in several PRs: make more code handle the unchecked index access without yet committing the TS config change.

**Comment by harrysolovay** [comment]
By "non-null" assertion I mean [the operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator).

```ts
declare const value: string | undefined
value! // `string`
```

We can apply these (`!`) pretty much everywhere there's a TS error stemming from potentially-undefined value access. I believe a large portion of these will need to be applied to the generated bindings of o1js-bindings. Might make sense to tweak the tsconfig within that directory alone.

**Comment by mitschabaude** [comment]
> By "non-null" assertion I mean [the operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator).

oh I see. I don't particularly like that operator, because it's neither type-safe nor a runtime assertion. It's like `as <the type without null>`. Using that operator everywhere is equivalent to not toggling `noUncheckedIndexAccess` on, which is fine as an upgrade path but not where I want to be.

> I believe a large portion of these will need to be applied to the generated bindings of o1js-bindings.

No, we don't generate any TS code apart from type definitions. Bindings artifacts are JS
